### DISCUSSION TITLE
Need help with a noise spectral density measurement using zhinst toolkit -- Missing path for Spectral Density option in Sweeper Module

### DISCUSSION THREAD

**Discussion by miguelairs** [zhinst/zhinst-toolkit#248] [Q&A] [answered]
Hello!
I am interested in using the zhinst python module and an MFLI Lock-in amplifier to do a noise spectral density measurement. I am very close to getting it (settings coded below), but there is one small difficulty. For a noise measurement in LabOne, I have to turn on the "spectral density" option screenshotted below. However, the LabOne API documentation does not provide any node for enabling spectral density calculations in the Sweeper Module. The other three LabOne options specify a path when I hover over them, but spectral density does not show any path at all. Does anyone know a fix for this? The code I used is largely based on the [sweeper example code](https://docs.zhinst.com/zhinst-toolkit/en/latest/examples/sweeper_module.html) and the [LabOne noise measurement info](https://www.zhinst.com/europe/en/blogs/noise-spectral-density-measured-lock-amplifiers).

If enabling the spectral density option in python is not possible, what are other ways to perform noise measurements using the zhinst-toolkit library? I know one from the LabOne page, but the formulation could be wrong (see very bottom). I appreciate any and all solutions and feedback!

Thanks so much!

![NoPath_SD](https://github.com/zhinst/zhinst-toolkit/assets/84990373/df5582ab-9dd4-4aeb-ad1f-11e8598b7fdc)
*No specified path for turning on spectral density*

![Path_Specified_Sinc](https://github.com/zhinst/zhinst-toolkit/assets/84990373/2169aba9-f58d-4e6f-8fc2-f0d9194fcfb6)
*Path specified for another option (sinc)*

Below is the code I am currently using:
```
# Connect the device
from zhinst.toolkit import Session
MFLI1_session = Session('192.168.99.218')
MFLI1 = MFLI1_session.connect_device("device_name")

# Configuring the instrument
OUT_CHANNEL = 0
OUT_MIXER_CHANNEL = 1 # UHFLI: 3, HF2LI: 6, MFLI: 1
IN_CHANNEL = 0
DEMOD_INDEX = 0
OSC_INDEX = 0


with MFLI1.set_transaction():
    MFLI1.sigins[IN_CHANNEL].ac(0)
    MFLI1.sigins[IN_CHANNEL].range(1)

    MFLI1.demods[DEMOD_INDEX].enable(True)
    MFLI1.demods[DEMOD_INDEX].rate(10e3)
    MFLI1.demods[DEMOD_INDEX].adcselect(IN_CHANNEL)
    MFLI1.demods[DEMOD_INDEX].order(4)
    MFLI1.demods[DEMOD_INDEX].timeconstant(0.01)
    MFLI1.demods[DEMOD_INDEX].oscselect(OSC_INDEX)
    MFLI1.demods[DEMOD_INDEX].harmonic(1)

    MFLI1.sigouts[OUT_CHANNEL].on(True)
    MFLI1.sigouts[OUT_CHANNEL].enables[OUT_MIXER_CHANNEL](1)
    MFLI1.sigouts[OUT_CHANNEL].range(1)
    MFLI1.sigouts[OUT_CHANNEL].amplitudes(OUT_MIXER_CHANNEL)


# Specify the number of sweeps to perform back-to-back.
LOOPCOUNT = 1

sweeper = MFLI1_session.modules.sweeper
sweeper.device(MFLI1)
```

Now we set the sweep module to perform the noise density measurement, with the exception of the missing "Spectral Density" activation. Most parameters are specified based on the [LabOne page for noise measurements](https://www.zhinst.com/europe/en/blogs/noise-spectral-density-measured-lock-amplifiers).

```
sweeper.gridnode(MFLI1.oscs[0].freq)
sweeper.start(10e3)
sweeper.stop(5e6) # 500e3 for MF devices, 50e6 for others
sweeper.samplecount(100)
sweeper.xmapping(1)
sweeper.filtermode(1)
sweeper.bandwidthcontrol(0)
sweeper.maxbandwidth(1.25e6)
sweeper.bandwidthoverlap(0)
sweeper.omegasuppression(60)
sweeper.scan(0)
sweeper.loopcount(LOOPCOUNT)
sweeper.settling.time(0)
sweeper.settling.inaccuracy(100e-9)
sweeper.settling.tc(24)
sweeper.averaging.tc(50)
sweeper.averaging.time(100e-3)
sweeper.averaging.sample(1000)

```

The rest follows the sweeper example from the official zhinst-toolkit page:

```
# Subscribing to sample node
sample_node = MFLI1.demods[0].sample
sweeper.subscribe(sample_node)

# Saving sweeper file
sweeper.save.filename('sweep_with_save')
sweeper.save.fileformat('hdf5')

# Execution
import logging
import sys
handler = logging.StreamHandler(sys.stdout)
logging.getLogger("zhinst.toolkit").setLevel(logging.INFO)
logging.getLogger("zhinst.toolkit").addHandler(handler)
sweeper.execute()
print(f"Perform {LOOPCOUNT} sweeps")
sweeper.wait_done(timeout=300)

sweeper.save.save(True)
# Wait until the save is complete. The saving is done asynchronously in the background
# so we need to wait until it is complete. In the case of the sweeper it is important
# to wait for completion before before performing the module read command. The sweeper has
# a special fast read command which could otherwise be executed before the saving has
# started.
sweeper.save.save.wait_for_state_change(True, invert=True, timeout=5)

# Read the data
data = sweeper.read()
sweeper.unsubscribe(sample_node)
```

Alternatively, the LabOne page explains the math behind the noise measurement. Using that formula one would go as follows:

```
import numpy as np

node_samples = data[sample_node]

for sample in node_samples:
    frequency = sample[0]["frequency"]
    xstddev = sample[0]["xstddev"]
    ystddev = sample[0]["ystddev"]
    bandwidth = sample[0]["bandwidth"]

noise = (xstddev**2 + ystddev**2)/(2*bandwidth)
```
If that's one of the alternative fixes, could someone verify that it's correct?

**Comment by mehdialem** [comment, marked as answer]
Dear @miguelairs,  
Thank you for your question. 

As you already observed, the "Spectral Density" is a post-processing method which is done in the graphical user interface. So it is not available as a node for the Sweeper module. Instead, you can simply calculate the spectral density according to the formula you already found in this [blog post](https://www.zhinst.com/blogs/noise-spectral-density-measured-lock-amplifiers). 

Your calculations at the end of your script seems to do the job. Please note that the obtained "noise" from your script is the power spectral density (PSD) in [V^2/Hz]. In case you are interested in the amplitude spectral density of noise in [V/rtHz], you just need to take the square root out of the calculated "noise" in your script.
